### DISCUSSION TITLE
How to reference utility classes added by plugins inside of `theme()` directive

### DISCUSSION THREAD

**Discussion by schardev** [tailwindlabs/tailwindcss#11126] [Help] [answered]
Suppose I have a plugin that adds utility classes for font sizes:

```js
module.exports = plugin.withOptions((opts) => {
  return ({ addUtilities }) => {
      const sizes = {
        "fluid-0": { fontSize: '<somevalue>' },
        "fluid-1": { fontSize: '<somevalue>' }
      }

     addUtilities(sizes)
  }
})
```

Now, I can reference them inside my css like:

```css
@layer components {
  .btn {
    @apply fluid-0;
  }
}
```

But how can I access them using the `theme()` directive?

```js
module.exports = {
  // ... other config
  extend: {
    typography: (theme) => ({
    DEFAULT: {
      css: {
        fontSize: theme(/* <how should I access them utilties here? */>)
      }
    }
    })
  }
}
```

Should I have to change something inside my plugin? Is there a way to add those utilities inside the default `fontSize` group in tailwind?

**Comment by wongjn** [comment]
You could look at creating ["dynamic" utility classes](https://tailwindcss.com/docs/plugins#dynamic-utilities), where the values are pulled from the `theme` configuration rather than hard-coded inside your plugin. This would mean you would be able to access them via the `theme()` function:

```js
module.exports = plugin(({ matchUtilities, theme }) => {
  matchUtilities(
    { fluid: value => ({ fontSize: value }) },
    { values: theme('fluidFontSizes') },
  );
});
```

```js
module.exports = {
  theme: {
    // Could also go in extend or be in the plugin's 
    // default values.
    fluidFontSizes: {
      1: '<somevalue>',
      2: '<somevalue>',
    },
    extend: {
      typography: ({ theme }) => ({
        DEFAULT: {
          css: {
            fontSize: theme('fluidFontSizes.1'),
          },
        },
      }),
    },
  },
  // …
};
```

**Reply by schardev** [reply]
But wouldn't that require the user to have static values inside their config? I don't expect users to have those values, so I'd rather want to "append" those utility classes to the default tailwind fontSize plugin. Is their a way to do that? 

I could add a default object to provide default values to the plugin but then I'd loose the ability to calculate dynamic values based on the options provided by user (as I have to calculate them inside the default `module.export` itself).

**Reply by wongjn** [reply]
> But wouldn't that require the user to have static values inside their config?

No, you could declare them as a plugin's default values.

> I'd rather want to "append" those utility classes to the default tailwind fontSize plugin. Is their a way to do that?

Yep! Add them to `theme.extend.fontSize` like:

```js
module.exports = {
  theme: {
    fontSize: {
      'fluid-1': '<somevalue>',
      'fluid-2': '<somevalue>',
    },
  },
  // …
};

**Reply by schardev** [reply, marked as answer]
@wongjn I think I misinterpreted your answer. I got it working like I wanted it to be with:

```ts
export = plugin.withOptions(
  () => {
    return ({ matchUtilities, theme }) => {
      matchUtilities(
        { text: (value: string) => ({ fontSize: value }) },
        { values: theme("fluidFontSize") }
      );
    };
  },
  (opts) => {
    const fontSizes = generateFontScales(opts);
    return {
      theme: { fluidFontSize: fontSizes },
    };
  }
);
```

I think the docs doesn't make the behaviour pretty clear but I got it right after tinkering with it.
